src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,101 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private T[] entryArray = (T[]) new Object[DEFAULT_CAPACITY];
+    private int size;
+
     @Override
     public void add(T value) {
+        extendArrayLength();
 
+        entryArray[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bounds");
+        }
+        extendArrayLength();
+        System.arraycopy(entryArray, index, entryArray, index + 1, size - index);
+        entryArray[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidity(index);
+        return (T) entryArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexValidity(index);
+        entryArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T element;
+        checkIndexValidity(index);
+        element = (T) entryArray[index];
+
+        arrayCopy(index);
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (entryArray[i] == null || entryArray[i].equals(element)) {
+                arrayCopy(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("No Such Element in array: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void extendArrayLength() {
+        if (entryArray.length <= size) {
+            int oldArrayLength = entryArray.length;
+            int growFactor = (int) (oldArrayLength * GROWTH_FACTOR); // extend array length by x1.5
```

Comment:
1. Variable name seems to be wrong - shouldn't this variable be called `newLength`?
2. no need to store `oldArrayLength` as variable, call `length` immediately here